pkg/controller/flink/client/error_handler.go

```diff
@@ -0,0 +1,160 @@
+package client
+
+import (
+	"fmt"
+	"math/rand"
+	"time"
+
+	"github.com/pkg/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/clock"
+)
+
+// appError codes
+const (
+	GlobalFailure      = "FAILED"
+	JSONUnmarshalError = "JSONUNMARSHALERROR"
+	defaultRetries     = 20
+	noRetries          = 0
+)
+
+type FlinkMethod string
+
+const (
+	CancelJobWithSavepoint FlinkMethod = "CancelJobWithSavepoint"
+	ForceCancelJob         FlinkMethod = "ForceCancelJob"
+	SubmitJob              FlinkMethod = "SubmitJob"
+	CheckSavepointStatus   FlinkMethod = "CheckSavepointStatus"
+	GetJobs                FlinkMethod = "GetJobs"
+	GetClusterOverview     FlinkMethod = "GetClusterOverview"
+	GetLatestCheckpoint    FlinkMethod = "GetLatestCheckpoint"
+	GetJobConfig           FlinkMethod = "GetJobConfig"
+	GetTaskManagers        FlinkMethod = "GetTaskManagers"
+	GetCheckpointCounts    FlinkMethod = "GetCheckpointCounts"
+	GetJobOverview         FlinkMethod = "GetJobOverview"
+)
+
+// FlinkApplicationError implements the error interface to make error handling more structured
+type FlinkApplicationError struct {
+	AppError            string       `json:"appError,omitempty"`
+	Method              FlinkMethod  `json:"method,omitempty"`
+	ErrorCode           string       `json:"errorCode,omitempty"`
+	IsRetryable         bool         `json:"isRetryable,omitempty"`
```

Comment:
What's the difference between IsRetryable and IsFailFast?

Reply:
A fail fast error implies that the operator will move the application to a deployFailed state in a single attempt (for example if we encounter a 400 in a submit job request). The idea here is that we'll have knowledge of what errors are definitely retryable/failfast (akin to retryable vs. throwable exceptions), so we specify them explicitly and everything else is handled on a wait-based recovery.

Reply:
I still don't quite understand why we have both IsRetryable and IsFailFast. Like, given this truth table:

|   | IsRetryable | IsFailFast |
|---|-------------|------------|
| 1 | T           | T          |
| 2 | T           | F          |
| 3 | F           | T          |
| 4 | F           | F          |

Which of those rows makes sense? It seems to me that only rows 2 and 3 make sense, so we can reduce this to a single variable.

Reply:
IMO 4) also makes sense, wherein when we encounter a class of errors that are [not](https://github.com/lyft/flinkk8soperator/blob/d4214227dd5a9c2b40eefb869e6f2510548e4b79/pkg/controller/flinkapplication/flink_state_machine.go#L636) necessarily from failing flink API calls (i.e. errors that we've not exhaustively enumerated), we default to a state of waiting for the application to make progress in the error-ed state. Alternatively, we could simplify all of this (as you suggested) and make aforementioned errors retryable by default.

Reply:
So in that case we're just waiting, but not retrying?

Reply:
Yes. That's right.

Reply:
Discussed offline and the approach we're going with is to only classify errors as retryable. Everything that's not retryable will be considered to "fail fast" i.e. move the application to a DeployFailed state in a single attempt.